IntegrationTests/Services/AWSS3IntegrationTests/SSOCredentialsProviderTests.swift

```diff
@@ -0,0 +1,116 @@
+    private var permissionSetArn: String = ""
+    private let awsReadOnlyPolicy = "arn:aws:iam::aws:policy/ReadOnlyAccess"
+    
+        
+    override func setUp() async throws {
+        try await super.setUp()
+        // Use default credentials provider chain for setup
+        ssoClient = try SSOAdminClient(region: "us-west-2")
+        try await createPermissionSetIfNeeded()
+        
+        // Create a S3 client that uses SSO credentials provider
+        try await setUpClient()
+    }
+
+    // The test calls listBuckets() and forces S3Client to use SSOCredentialsProvider
+    func test_listBuckets() async throws {
+        _ = try await client.listBuckets(input: ListBucketsInput())
+    }
+    
+    /* HELPER METHODS */
+    private func createPermissionSetIfNeeded() async throws {
+        // Get IAM identity center instanceArn
+        let listInstancesOutput = try await ssoClient.listInstances(input: ListInstancesInput())
+        guard let iamCenterInstances = listInstancesOutput.instances, let iamCenterArn = iamCenterInstances[0].instanceArn else {
+            throw ClientError.dataNotFound("No IAM Identity Center instance found for the account."
+                             + "Ensure that AWS organization is enabled for the account you're using to run the integration test.")
+        }
+        iamIdentityCenterInstanceArn = iamCenterArn
+        
+        do {
+            // Create permission set and attach AWS managed read-only policy
+            try await createPermissionSet()
+        } catch let error as AWSSSOAdmin.ConflictException {
+            // Catch error if permission set has already been created from previous run of this integ test
+            if let message = error.message, message.contains("\(permissionSetName) already exists") {
+                return
+            } else {
+                throw error
+            }
+        }
+    }
+    
+    private func createPermissionSet() async throws {
+        // Create permission set and save its ARN
+        let createPermissionSetOutput = try await ssoClient.createPermissionSet(input: CreatePermissionSetInput(
+            description: "Permission set for testing SSO credentials provider.",
+            instanceArn: iamIdentityCenterInstanceArn,
+            name: permissionSetName
+        ))
+        
+        guard let permSet = createPermissionSetOutput.permissionSet, let permSetArn = permSet.permissionSetArn else {
+            throw ClientError.dataNotFound("Permission set arn could not be retrieved after creation.")
+        }
+        permissionSetArn = permSetArn
+        
+        // Attach ReadOnly AWS-managed policy to the created permission set
+        _ = try await ssoClient.attachManagedPolicyToPermissionSet(input: AttachManagedPolicyToPermissionSetInput(
+            instanceArn: iamIdentityCenterInstanceArn,
+            managedPolicyArn: awsReadOnlyPolicy,
+            permissionSetArn: permissionSetArn))
```

Comment:
I think we can make it even more readable. Can we make 2 helper functions here. Maybe something like

`getPermissionSet()` which does
```
let input = CreatePermissionSetInput(
    description: "Permission set for testing SSO credentials provider.",
     instanceArn: iamIdentityCenterInstanceArn,
     name: permissionSetName
)
  
let output = try await ssoClient.createPermissionSet(input: input)
guard let permSetArn = output.permissionSet?.permissionSetArn else {
    throw ClientError.dataNotFound("Permission set ARN not found after creation.")
}
return permSetArn
```

and  then

`attachManagedPolicy(arn: String)` which does
```
let input = AttachManagedPolicyToPermissionSetInput(
      instanceArn: iamIdentityCenterInstanceArn,
      managedPolicyArn: awsReadOnlyPolicy,
      permissionSetArn: arn
)
_ = try await ssoClient.attachManagedPolicyToPermissionSet(input: input)
```

then `createAndAttachPermissionSet()` instead of `createPermissionSet()` can do 
```
// create permission set and return arn
let permSetArn = try await getPermissionSet()
// attach created policy
try await attachManagedPolicy(arn: permSetArn)
```

Reply:
Refactored to smaller scope for each function as suggested - added three functions.